main/autogen/constant_hash.cc

```diff
@@ -0,0 +1,125 @@
+#include "main/autogen/constant_hash.h"
+#include "common/typecase.h"
+#include "core/hashing/hashing.h"
+#include "parser/Node.h"
+
+using namespace std;
+namespace sorbet::autogen {
+
+// See the comment in `main/autogen/constant_hash.h` for the intention
+// and high-level behavior of `constantHashNode`.
+//
+// A few operational notes about this function:
+//
+// - the reason it takes a parsed AST (rather than, say, a desugared
+// one) is that this means we can do less allocation and run the hash
+// as quickly as possible.
+//
+// - the reason it takes a pointer rather than a `unique_ptr` is so it
+// doesn't need to `move` or consume the AST: it just observes it. The
+// reason it takes a pointer rather than a reference is so we don't
+// need to check for nullability on every recursive call: we can check
+// once at the head of `hashNode` instead.
```

Comment:
Do you have an objection to taking `const unique_ptr<Node> &`?  That would at least let you get rid of the `.get()` calls and should be roughly as efficient.

Reply:
None at all! That's easy enough to do.

Reply:
++ on this change